tools/nl/embeddings/finetune.py

```diff
@@ -0,0 +1,258 @@
+
+  # Now add the manual sentence pairs (which are assumed to have the scores).
+  for _, row in df_sentence_pairs.iterrows():
+    assert "sentence_1" in row
+    assert "sentence_2" in row
+    assert "score" in row
+    training_examples.append(
+        InputExample(texts=[row["sentence_1"], row["sentence_2"]],
+                     label=row["score"]))
+
+  # Ready to fine-tune.
+  train_dataloader = DataLoader(training_examples,
+                                shuffle=True,
+                                batch_size=BATCH_SIZE)
+  t = time.time()
+  model.fit(train_objectives=[(train_dataloader,
+                               losses.CosineSimilarityLoss(model=model))],
+            epochs=NUM_EPOCHS,
+            warmup_steps=NUM_WARMUP_STEPS)
+  t = time.time() - t
+  print(f"Time taken = {t}")
+  return model
+
+
+def main(_):
+
+  assert FLAGS.model_name_v2 and FLAGS.bucket_name_v2
+  assert FLAGS.embeddings_size in ['small', 'medium']
+
+  assert os.path.exists(os.path.join('data'))
+
+  autogen_input_filepattern = f'{FLAGS.autogen_input_basedir}/{FLAGS.embeddings_size}/*.csv'
+
+  gs = gspread.oauth()
+  sc = storage.Client()
+  bucket = sc.bucket(FLAGS.bucket_name_v2)
+
+  df_svs = _alternatives(autogen_input_filepattern,
+                         FLAGS.alternatives_filepattern)
+  df_sentence_pairs = pd.read_csv(FLAGS.sentence_pairs_filepath).fillna("")
+  model = SentenceTransformer(FLAGS.model_name_v2)
+  model = fine_tune_model(model, df_svs, df_sentence_pairs)
+
+  ctx = utils.Context(gs=gs, model=model, bucket=bucket, tmp='/tmp')
+
+  gcs_model_folder = _make_gcs_model_folder(FLAGS.model_name_v2)
+  gcs_tmp_out_path = os.path.join(ctx.tmp, gcs_model_folder)
+
+  print(f"Saving locally to {gcs_tmp_out_path}")
+  model.save(gcs_tmp_out_path)
+
+  # Finally, upload to the NL model server's GCS bucket
+  print("Attempting to write to GCS")
+  print(f"\t GCS Path: gs://{FLAGS.bucket_name_v2}/{gcs_model_folder}/")
+
+  # To upload the model directory, we need to traverse the files and folders.
+  for str_path in glob.glob(f"{gcs_tmp_out_path}/**"):
+
+    # Check if str_path is a folder.
+    if os.path.isdir(str_path):
+      if not glob.glob(f"{str_path}/**"):
+        # This means we found an empty folder.
+        foldername = os.path.basename(str_path)
+        gcs_path = gcs_model_folder + "/" + foldername + "/"
+        print(f'Path in GCS: {gcs_path}')
+        _upload_to_gcs(ctx, gcs_path, str_path, empty_folder=True)
+
+      for filepath in glob.glob(f"{str_path}/**"):
+        # Found files under a folder.
+        filename = filepath.split(gcs_tmp_out_path)[1]
+        gcs_path = gcs_model_folder + filename
+        _upload_to_gcs(ctx, gcs_path, filepath)
+    else:
+      # Just files under the main model folder.
+      foldername = os.path.basename(str_path)
+      gcs_path = gcs_model_folder + "/" + foldername
+      _upload_to_gcs(ctx, gcs_path, str_path)
```

Comment:
nit: could this perhaps be in a save_model function?

so the top-level function does: loading, fine tuning, saving

Reply:
Done